Pre-match Diagnostics Testing

Ace5tar opened the pull request.

Ace5tar committed:
fix?

Ace5tar committed:
fix?

Ace5tar committed:
drivebase test command & pivot test command

Ace5tar committed:
drivebase test command & pivot test command

Ace5tar committed:
test shooter + indexer implementation

Ace5tar committed:
test shooter + indexer implementation

witherrrr commented on src/main/java/frc/robot/Diagnostics.java:
```diff
@@ -0,0 +1,39 @@
+  public static Command testIndexerCommand(Indexer indexer) {
+    return indexer.indexerOnIntakeOnCommand()
+    .andThen(indexer.indexerOffIntakeOffCommand());
+  }
```
this will run both of them for one code cycle (20ms) and then immediately stop because the commands are runOnce, you should add a wait command in between

witherrrr replied:
nvm all of indexer is bugged lol, can you change all of the `runOnce` commands in `Indexer` to be `run` and then you can use `command.withTimeout` to run for a specific period of time

the issue with `runOnce` is that the command will run for one code cycle but then the default command will take over again, and we want to have control over the system operation for the required time

witherrrr reviewed (commented):
all of the diagnostic commands look good, next step is figure out how to run these in the pit (test mode? auto routine that does them all one after the other? etc)

Ace5tar committed:
Merge branch 'Diagnotics' of https://github.com/AEMBOT/FRC_2024 into Diagnotics

Ace5tar committed:
pre-match test commands

nolanhergert reviewed (commented):
Looking good! Looking forward to seeing this run!

Ace5tar committed:
Merge branch 'main' of https://github.com/AEMBOT/FRC_2024 into Diagnotics

Ace5tar committed:
reworked diagnotic system+implemented climber test

Ace5tar committed:
Merge branch 'main' of https://github.com/AEMBOT/FRC_2024 into Diagnotics

Ace5tar committed:
minor diag changes + individual drive motor tests

Ace5tar committed:
individual climber motor testing

Ace5tar committed:
Merge branch 'main' of https://github.com/AEMBOT/FRC_2024 into Diagnotics

nolanhergert commented on src/main/java/frc/robot/Diagnostics.java:
```diff
@@ -0,0 +1,162 @@
+  // wibble wobble switch between two commands for 1 second
+  private static Command testWobble(Command minCommand, Command maxCommand) {
+    return repeatingSequence(
+            minCommand.withTimeout(0.25), WaitCommand(0.25), maxCommand.withTimeout(0.25))
```
You'll want these 0.25's as a constant so it's easier to make sure they are the same. "No magic numbers" is the refrain.

private static final double wobbleTimeout = 0.25;
minCommand.withTimeout(wobbleTimeout), etc.

nolanhergert commented on src/main/java/frc/robot/Diagnostics.java:
```diff
@@ -0,0 +1,162 @@
+      driveCommand =
+          driveCommand
+              .andThen(drive.singleMotorDriveTest(modules[i], 4))
+              .withTimeout(1)
```
Same for these timeouts and voltages.

nolanhergert commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -219,7 +219,66 @@ public RobotContainer() {
+                            .withName("sub"))),
+                Commands.sequence(
+                    runOnce(() -> Logger.recordOutput("autoState", 0.04)),
+                    new ProxyCommand(
+                        shooter.setVelocityRPMCommand(shooterSpeedRPM).withName("shoot")))),
+            runOnce(() -> Logger.recordOutput("autoState", 0.3)),
+            Commands.sequence(runOnce(() -> Logger.recordOutput("autoState", 0.4))),
+            new ProxyCommand(
+                pivot.setPositionCommand(() -> Units.degreesToRadians(40)).withTimeout(0.3))));
+
+    autoChooser.addOption(
+        "REAL FOUR PIECE",
+        Commands.sequence(
+            runOnce(() -> Logger.recordOutput("autoState", -1)),
+            Commands.deadline(
+                    Commands.sequence(
+                        waitSeconds(0.2),
+                        runOnce(() -> Logger.recordOutput("autoState", 0.1)),
+                        waitUntil(() -> pivot.atGoal() && shooter.isAtShootSpeed())
+                            .withTimeout(0.3),
+                        runOnce(() -> Logger.recordOutput("autoState", 0.2)),
+                        new ProxyCommand(indexer.shootCommand().withTimeout(0.3))),
+                    pivot.setPositionCommand(() -> Units.degreesToRadians(60)),
+                    shooter.setVelocityRPMCommand(shooterSpeedRPM))
+                .andThen(
+                    Commands.sequence(runOnce(() -> Logger.recordOutput("autoState", 0.3))),
+                    pivot.setPositionCommand(() -> Units.degreesToRadians(40)).withTimeout(0.3)),
+            runOnce(() -> Logger.recordOutput("autoState", 1)),
+            AutoBuilder.followPath(PathPlannerPath.fromChoreoTrajectory("four-piece-amp.1")),
+            runOnce(() -> Logger.recordOutput("autoState", 2)),
+            Commands.race(Commands.waitSeconds(1), NamedCommands.getCommand("shootNoteAuto")),
+            runOnce(() -> Logger.recordOutput("autoState", 3)),
+            AutoBuilder.followPath(PathPlannerPath.fromChoreoTrajectory("four-piece-amp.2")),
+            runOnce(() -> Logger.recordOutput("autoState", 4)),
+            Commands.race(Commands.waitSeconds(1), NamedCommands.getCommand("shootNoteAuto")),
+            runOnce(() -> Logger.recordOutput("autoState", 5)),
+            AutoBuilder.followPath(PathPlannerPath.fromChoreoTrajectory("four-piece-amp.3")),
+            runOnce(() -> Logger.recordOutput("autoState", 6)),
+            Commands.race(Commands.waitSeconds(1), NamedCommands.getCommand("shootNoteAuto")),
+            runOnce(() -> Logger.recordOutput("autoState", 7))));
```
I think you accidentally included this when merging in main or something. I can help you revert this in git extensions

nolanhergert commented on src/main/java/frc/robot/Diagnostics.java:
```diff
@@ -0,0 +1,162 @@
+    }
+    return driveCommand;
+  }
+
```
After talking with Deryk, he suggested commanding all modules to go in the same orientation and the same speed so you can make sure the modules are aligned still.

After talking with James, it sounds like it'd be good to do an "around-the-world" rotation type thing, they just had a bug with that.

Would be something like this I think.
drive.runVelocity(new ChassisSpeeds(0.1, 0, 0.1));

nolanhergert commented on src/main/java/frc/robot/Diagnostics.java:
```diff
@@ -0,0 +1,162 @@
+    double avgClimberHeight = (minClimberHeight + maxClimberHeight) / 2;
+
+    return climber
+        .setLeftPositionCommand(minClimberHeight)
```
You might need to home first. Not sure

nolanhergert commented on src/main/java/frc/robot/Diagnostics.java:
```diff
@@ -0,0 +1,162 @@
+
+    return climber
+        .setLeftPositionCommand(minClimberHeight)
+        .withTimeout(1)
```
Do you want it to go to the minimum and maximum? If so, you don't want a timeout and instead wait for it to finish.

The "both" command does it properly:
```
  public Command setPositionCommand(double targetPosition) {
    return run(() -> setPosition(targetPosition)).finallyDo(() -> io.setVoltage(0.0));
  }
```

I don't think you need to run the climbers separately actually. It's pretty easy to tell them apart auditorially and visually. We might even want to compare the velocities visually too!

nolanhergert commented on src/main/java/frc/robot/subsystems/climber/ClimberIOSparkMax.java:
```diff
@@ -128,11 +128,29 @@ public void setLeftVoltage(double volts) {
     m_winchMotorLeft.setVoltage(volts);
   }
 
+  public void setLeftPosition(double position) {
```
Won't need these imo

nolanhergert commented on src/main/java/frc/robot/subsystems/climber/Climber.java:
```diff
@@ -83,4 +83,12 @@ public double getCurrentLimit(boolean homingBool) {
     }
   }
+
+  public Command setLeftPositionCommand(double position) {
```
I don't think you will need these, based on my command above.

nolanhergert commented on src/main/java/frc/robot/subsystems/indexer/Indexer.java:
```diff
@@ -148,4 +148,12 @@ public Command indexerOutCommand() {
     return runOnce(this::indexerStop);
   }
+
+  public Command indexerOnIntakeOnCommand() {
```
I don't think you need these. Try to copy how the commands are already done from the RobotContainer as much as possible so you can make sure those flows work still (beam brake, etc)

// Intake Manual In
    controller.rightBumper().whileTrue(indexer.getDefault(pivot::inHandoffZone));

Which calls this:
 run(this::indexOffIntakeOn)
                        .until(() -> inputs.intakeBeamBreakState)
                        .andThen(run(this::indexOffIntakeOff))
                        .until(pivotHandoff)
                        .andThen(run(this::indexOnIntakeOn))))

nolanhergert committed:
Only tested pivot, which works well along with Wobble!

nolanhergert committed:
For some reason running the test command from Driver station didn't advance in pivot test.

We need to figure that out. Workaround is to hold the left bumper of backup controller.

Also don't test other modules until they are ready

Ace5tar committed:
I did a lot so im listing them below this
added some more constants and moved all of the constants into Constants.java
added an "around the world" test to drive train
reworked indexer testing to use commands that pre-existed
also maybe didnt push april tag testing code yesterday (?) i thought i did but looks like its getting pushed with this commit

Ace5tar commented:
This pr has been open for so long and will almost certainly never get merged, so I'm going to close it.
